### ISSUE TITLE
ch 8 Getting 48% accuracy in the 'Working with bigger data- online algorithms and out of core learning'

### ISSUE THREAD

**Issue by LasaniHussain** [rasbt/python-machine-learning-book-2nd-edition#67]
I typed in the same code as in the book and all functions : get_minbatch(),stream_docs(). The code runs correctly but gives 48% accuracy.I also have checked and confirmed that I am using 45000 reviews for traininig and rest 5000 reviews for testing.Please help as I have to further embed it in a web app.

**Comment by LasaniHussain** [comment]
SIr, I wrote the code over and over again, double checked it from the jupyter code hosted on github but still it is giving 48% accuracy.Also the code in your book takes 38 seconds but the same code on my laptop takes 9 seconds to run as reflected by pyprind progress bar, though I have no GPU.I thought that it must not be training on the whole data but it does train for 45 iterations processing 1000 reviews in each. 
Also I am feeling great to get reply from you.You made understanding machine learning algorithms so easy, even easier than any online MOOC that I took.I have done one research conference on Energy Load forecasting using Spark cluster which got accepted in IEEE though I am an undergraduate student from a tier 3 college in India.I owe you a lot.

**Comment by LasaniHussain** [comment]
Sorry I clicked the close by mistake

**Comment by LasaniHussain** [comment]
SIr, I wrote the code over and over again, double checked it from the jupyter code hosted on github but still it is giving 48% accuracy.Also the code in your book takes 38 seconds but the same code on my laptop takes 9 seconds to run as reflected by pyprind progress bar, though I have no GPU.I thought that it must not be training on the whole data but it does train for 45 iterations processing 1000 reviews in each.
Also I am feeling great to get reply from you.You made understanding machine learning algorithms so easy, even easier than any online MOOC that I took.I have done one research conference on Energy Load forecasting using Spark cluster which got accepted in IEEE though I am an undergraduate student from a tier 3 college in India.I owe you a lot.

**Comment by rasbt** [comment]
Thanks for the very kind words :). I am happy to hear that you found the book useful, and congrats on the IEEE paper, that's awesome and a big accomplishment!

> Also the code in your book takes 38 seconds but the same code on my laptop takes 9 seconds to run as reflected by pyprind progress bar, though I have no GPU

It shouldn't depend on whether you have a GPU because it runs on the CPU. I just reran the code and it takes about 30 seconds  on my laptop. 
I have a suspicion that the `movie_data.csv` file may not be complete. Could you check the file size? It should be roughly 66 Mb.

**Comment by LasaniHussain** [comment]
Sir, I checked the movie_data.csv file and it is 65.9 mb

**Comment by rasbt** [comment]
Hm, could you maybe try to run the following code as a script, which is an abbreviated version of the code in the book (e.g., via `python3 test_ch08.py` from your command line):

```python
import re
import pyprind
import numpy as np
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.linear_model import SGDClassifier

nltk.download('stopwords')
stop = stopwords.words('english')


def tokenizer(text):
    text = re.sub('<[^>]*>', '', text)
    emoticons = re.findall('(?::|;|=)(?:-)?(?:\)|\(|D|P)', text.lower())
    text = re.sub('[\W]+', ' ', text.lower()) + ' '.join(emoticons).replace('-', '')
    tokenized = [w for w in text.split() if w not in stop]
    return tokenized


def stream_docs(path):
    with open(path, 'r', encoding='utf-8') as csv:
        next(csv)  # skip header
        for line in csv:
            text, label = line[:-3], int(line[-2])
            yield text, label


next(stream_docs(path='movie_data.csv'))


def get_minibatch(doc_stream, size):
    docs, y = [], []
    try:
        for _ in range(size):
            text, label = next(doc_stream)
            docs.append(text)
            y.append(label)
    except StopIteration:
        return None, None
    return docs, y


vect = HashingVectorizer(decode_error='ignore',
                         n_features=2**21,
                         preprocessor=None,
                         tokenizer=tokenizer)

clf = SGDClassifier(loss='log', random_state=1)
doc_stream = stream_docs(path='movie_data.csv')

pbar = pyprind.ProgBar(45)

classes = np.array([0, 1])
for _ in range(45):
    X_train, y_train = get_minibatch(doc_stream, size=1000)
    if not X_train:
        break
    X_train = vect.transform(X_train)
    clf.partial_fit(X_train, y_train, classes=classes)
    pbar.update()


X_test, y_test = get_minibatch(doc_stream, size=5000)
X_test = vect.transform(X_test)
print('Accuracy: %.3f' % clf.score(X_test, y_test))

clf = clf.partial_fit(X_test, y_test)
```

On my machine, I get

    Total time elapsed: 00:00:25
    Accuracy: 0.867

**Comment by LasaniHussain** [comment]
Sir on running the script I got .879 accuracy but why was it not working in
the Jupyter notebook, I couldn't unserstand.


On Tue, Apr 24, 2018 at 9:47 PM, Sebastian Raschka <notifications@github.com

**Comment by rasbt** [comment]
I am closing this for now, please let me know if you need any further help with that
